### DISCUSSION TITLE
Cart Transform API bug - customized bundles

### DISCUSSION THREAD

**Discussion by meandillar** [Shopify/function-examples#378] [Feedback: Cart Transform API]
I've been experimenting with Cart Transform functions and came across some unusual behaviour when using different combinations of Merge and Expand operations and switching between cart and checkout.

Let's say I have 2 bundle configurations:

1. an Expand bundle with the following configuration
Parent ID: `gid://shopify/ProductVariant/111`
Child Component ID:`gid://shopify/ProductVariant/211` (caramel)
Child Component ID: `gid://shopify/ProductVariant/211` (caramel)

2. a Merge bundle
Parent variant ID: `gid://shopify/ProductVariant/111`
Child Component ID: `gid://shopify/ProductVariant/212` (chocolate)
Child Component ID: `gid://shopify/ProductVariant/212` (chocolate)

Things to note: 
- both parents are the same for each bundle
- the child components of both bundles come from the same product but are different flavours
- the parents and children are from different products
- both bundles work when only one configuration or the other is added to metafields

**Behaviour:**
If you add 2 x `gid://shopify/ProductVariant/211` (caramel) to the cart, it shows 1 x `gid://shopify/ProductVariant/111` (parent) in the cart. Upon navigation to checkout it shows the 2 x caramel variants nested, which is correct. 

Navigate back to the store and add 2 x `gid://shopify/ProductVariant/212` (chocolate) and it shows 2 x `gid://shopify/ProductVariant/111` (parent) in the cart. Upon navigation to checkout it shows 4 x `gid://shopify/ProductVariant/212` (chocolate) nested beneath the parent.

**Expected Checkout:**
- A bundle with 2 x `gid://shopify/ProductVariant/212` (chocolate)
- A bundle with 2 x `gid://shopify/ProductVariant/211` (caramel)

I can assume from this behaviour that there is either a bug or there is some kind of limitation where you can't associate multiple bundle configurations with a single variant.

**Comment by horst7000** [comment]
I experience a very similar issue when using the same parentVariant for multiple customized bundles, except I only use merge operations. 

I could see a possible solution where not only the **parentVariantId** but also the **title** gets taken into consideration before merging effectively different bundles together and keep them separate cart items instead.

**Comment by meandillar** [comment]
@horst7000 I've tested using only merge operations targeting the same parent variant with different child component variants and I'm also experiencing bugs.

I'm wondering if the only way around this problem is to create a new parent variant for each bundle possibility. This is not ideal though because when bundling products that have several options/variants we will need to create lots of variants. The ideal scenario for us would be having a single parent variant and being able to select the child component variants from a product selector to attach to that parent dynamically.

**Comment by Ale1234509** [comment]
Any update regarding this one? I'm experiencing the same issues with the merge operation.

**Comment by anthony-v-sundae-lane** [comment]
Confirming we're experiencing the same issue. I flagged this in a recent AMA on Discord and didn't get a response beyond the initial saying its supposed to work (ie, create two bundles, not a single combined bundle)

**Comment by rnovacek** [comment]
Any update on this? This is the biggest blocker for us that prevents us from using native bundles.

**Comment by nickwesselman** [comment]
Converted to a discussion as this appears to be a feature request.

**Reply by meandillar** [reply]
@nickwesselman Seeing there is no documentation regarding whether or not it is possible to associate multiple bundles to a single parent variant this is not really a feature request but more of a bug report. The behaviour of items changing between cart and checkout is unexpected.

It is deeply discouraging that issues like this get relegated to "discussions" and then are seemingly overlooked and forgotten about.

**Reply by rnovacek** [reply]
My thoughts exactly. It's undocumented behaviour and we cannot tell if it's intended or just a bug.

For me, it's bug in any case - the question is whether it is a bug in the code (multiple bundles with the same parent should be working) or bug in the documentation (it's not supposed to work but the documentation fails to explain so).

**Comment by derekmorash** [comment]
Running into the same problem trying to create custom bundles by merging multiple items into a parent. Works fine until you add more than one bundle with different child products because in the cart it's seen as the same parent item and groups them under the same cart line item. Make's it basically unusable for custom bundles unfortunately.

Would love to hear any potential solutions or any plans to address this, feels like an incomplete feature right now.

**Comment by siddhantbajaj** [comment]
Update - Sorry for the delayed response but we have identified the issue and are working on a fix. I'll keep you posted here once the fix is merged.

**Comment by siddhantbajaj** [comment]
We made some changes, can folks here confirm if they are still facing this issue? If so can you share the steps to replicate, along with your development store url?

**Reply by rnovacek** [reply]
Thank you, it seems that the fix has resolved our problems. Multiple bundles (containing different products) can be now added to cart using the same parent variant.

**Reply by anthony-v-sundae-lane** [reply]
Can confirm its now working as expected on our end. Thanks!

Is it now referencing the Title to signify unique bundles?

**Reply by stevekaran** [reply]
In my testing, I'm experiencing a similar issue with updating quantities. I'm able to add more than one of the same parent bundle to the cart. provided the bundle is comprised of different components (i.e. bundle_a: flavour1, flavour2; bundle_a, flavour3, flavour4). However, when I increase the quantity of a bundle (using /cart/update.js), the quantity of the components is updated, but not the parent bundle, as illustrated by this checkout. Doesn't it make more sense for the bundle parent to be quantity 2 as well?

![image](https://github.com/Shopify/function-examples/assets/149213887/b26af872-b360-4e54-91f7-7124c924ed2e)

**Reply by Andscowal** [reply]
Hi @siddhantbajaj, it seems that this is still an issue per the note above from back in November along with more recent parent level comments. We're experiencing this exact problem still as well. Has there been any movement on a resolution?

**Comment by SaeedGHF** [comment]
We have the same issue when multiple number of the same bundle exists in cart.

After navigating to checkout the price update modal is shown and child components are added back to cart.

Example output:
`{
  "operations": [
    {
      "merge": {
        "cartLines": [
          {
            "cartLineId": "gid://shopify/CartLine/075d4d14-4bf2-4389-b4e4-f7886cd7bb2e",
            "quantity": 1
          },
          {
            "cartLineId": "gid://shopify/CartLine/2a56590c-acd7-4595-adfe-a80b11994646",
            "quantity": 1
          }
        ],
        "parentVariantId": "gid://shopify/ProductVariant/47599165669668",
        "price": {
          "percentageDecrease": {
            "value": 25
          }
        },
        "image": null,
        "title": "1"
      }
    },
    {
      "merge": {
        "cartLines": [
          {
            "cartLineId": "gid://shopify/CartLine/075d4d14-4bf2-4389-b4e4-f7886cd7bb2e",
            "quantity": 1
          },
          {
            "cartLineId": "gid://shopify/CartLine/2a56590c-acd7-4595-adfe-a80b11994646",
            "quantity": 1
          }
        ],
        "parentVariantId": "gid://shopify/ProductVariant/47599165669668",
        "price": {
          "percentageDecrease": {
            "value": 25
          }
        },
        "image": null,
        "title": "2"
      }
    },
    {
      "merge": {
        "cartLines": [
          {
            "cartLineId": "gid://shopify/CartLine/075d4d14-4bf2-4389-b4e4-f7886cd7bb2e",
            "quantity": 1
          },
          {
            "cartLineId": "gid://shopify/CartLine/2a56590c-acd7-4595-adfe-a80b11994646",
            "quantity": 1
          }
        ],
        "parentVariantId": "gid://shopify/ProductVariant/47599165669668",
        "price": {
          "percentageDecrease": {
            "value": 25
          }
        },
        "image": null,
        "title": "3"
      }
    },
    {
      "merge": {
        "cartLines": [
          {
            "cartLineId": "gid://shopify/CartLine/075d4d14-4bf2-4389-b4e4-f7886cd7bb2e",
            "quantity": 1
          },
          {
            "cartLineId": "gid://shopify/CartLine/2a56590c-acd7-4595-adfe-a80b11994646",
            "quantity": 1
          }
        ],
        "parentVariantId": "gid://shopify/ProductVariant/47599165669668",
        "price": {
          "percentageDecrease": {
            "value": 25
          }
        },
        "image": null,
        "title": "4"
      }
    }
  ]
}`

**Reply by SaeedGHF** [reply]
The issue also exists on the same operation title.

Example output:
`{
  "operations": [
    {
      "merge": {
        "cartLines": [
          {
            "cartLineId": "gid://shopify/CartLine/9dff30c8-b084-4e53-ab6c-38b1d0ee5d0a",
            "quantity": 1
          },
          {
            "cartLineId": "gid://shopify/CartLine/96dbc3d2-73b6-4544-bf27-3c434f74d43f",
            "quantity": 1
          }
        ],
        "parentVariantId": "gid://shopify/ProductVariant/47599165669668",
        "price": {
          "percentageDecrease": {
            "value": 25
          }
        },
        "image": null,
        "title": "Bundle 34"
      }
    },
    {
      "merge": {
        "cartLines": [
          {
            "cartLineId": "gid://shopify/CartLine/9dff30c8-b084-4e53-ab6c-38b1d0ee5d0a",
            "quantity": 1
          },
          {
            "cartLineId": "gid://shopify/CartLine/96dbc3d2-73b6-4544-bf27-3c434f74d43f",
            "quantity": 1
          }
        ],
        "parentVariantId": "gid://shopify/ProductVariant/47599165669668",
        "price": {
          "percentageDecrease": {
            "value": 25
          }
        },
        "image": null,
        "title": "Bundle 34"
      }
    }
  ]
}`

**Reply by ashfaqhaq** [reply]
facing the same issue, would love to hear if you were able to find the resolution

**Reply by SaeedGHF** [reply]
Unfortunately, I have not been able to find a solution to this problem yet.

**Comment by ashfaqhaq** [comment]
Facing the same issues as Saeed mentioned,

Once we add more than one quantity of bundles in cart, then shopify adds all other items to the cart as individual products and it messes up everything..
![image](https://github.com/Shopify/function-examples/assets/44408323/a3919fa1-5c4b-41c8-9e79-f3e0ddb740c7)

**Comment by rnovacek** [comment]
We are having an issue that is quite similar:

When I add products (= one bundle) into cart, the cart transform is applied correctly. But if I go back to the cart page, remove that bundle, then create a new one with different products (but the same ParentVariant), it's fine on the cart page, but in the checkout I still see the old bundle. It's like the checkout page is cached but shouldn't be.

Also it breaks when I change the "bundle" product quantity on cart page. It will split it into individual products instead of showing it as a bundle.

@siddhantbajaj Are you guys aware of these issues? Any ETA when it might be fixed? We have already launched "native bundles" live to the merchants and now we're getting bad reviews because of issues like that what we're not in control of.

**Comment by spinturnix** [comment]
I'm experiencing the same issue as @stevekaran.

I have a working merge function, but after increasing quantity in the cart it'll increase component quantities while the parent always remains at 1. I'd rather have the opposite of this, where only the parent qty is changing. Though when I define a specific item count for the merge, increasing quantity greater than 1 breaks the components out into their own lines again.

Do I need to return multiple identical merge operations to get the correct quantity?

@nickwesselman @siddhantbajaj 🙏

**Comment by Hujjat** [comment]
Having same issue. 

https://github.com/Shopify/function-examples/assets/5115878/b815a000-2d52-423e-8d1d-2d3506d1b62b

Also, the merge operation doesn't show discounted price. It's very important. because we offer discount when customer bundle products.

**Reply by nelson-maze** [reply]
Confirming this behavior is still happening in a Plus Sandbox environment with similar behavior.

Updating the quantity of the parent in the cart continues to result in some odd duplication issues.
